cpp/src/neighbors/ivf_flat/jit_lto_kernels/generate_kernels.py

```diff
@@ -0,0 +1,470 @@
+
+#!/usr/bin/env python3
+"""
+Generate CUDA kernel instantiation files for IVF-Flat interleaved scan.
+This script generates kernel files programmatically based on type combinations.
+"""
+
+from pathlib import Path
+import itertools
+
+
+# Define the parameter space for kernel generation
+CAPACITIES = [0, 1, 2, 4, 8, 16, 32, 64, 128, 256]
+ASCENDING_VALUES = [True, False]
+COMPUTE_NORM_VALUES = [True, False]
+
+# Data type configurations: (data_type, acc_type, veclens, type_abbrev, acc_abbrev)
+# Each data type has veclen=1 and one optimized larger veclen
+DATA_TYPE_CONFIGS = [
+    ('float', 'float', [1, 4], 'f', 'f'),
+    ('__half', '__half', [1, 8], 'h', 'h'),
+    ('uint8_t', 'uint32_t', [1, 16], 'uc', 'ui'),
+    ('int8_t', 'int32_t', [1, 16], 'sc', 'i'),
+]
+
+IDX_TYPE = 'int64_t'
+IDX_TYPE_ABBREV = 'l'
+
+# Metric configurations for device functions
+METRIC_CONFIGS = [
+    'euclidean',
+    'inner_prod',
+]
+
+# Filter configurations
+FILTER_CONFIGS = [
+    'filter_none',
+    'filter_bitset',
+]
+
+# Post lambda configurations
+POST_LAMBDA_CONFIGS = [
+    'post_identity',
+    'post_sqrt',
+    'post_compose',
+]
+
+
+def generate_kernel_combinations():
+    """Generate all valid kernel parameter combinations."""
+    kernels = []
+
+    for data_type, acc_type, veclens, type_abbrev, acc_abbrev in DATA_TYPE_CONFIGS:
+        for capacity, veclen, ascending, compute_norm in itertools.product(
+            CAPACITIES, veclens, ASCENDING_VALUES, COMPUTE_NORM_VALUES
+        ):
+            kernels.append({
+                'capacity': capacity,
+                'veclen': veclen,
+                'ascending': ascending,
+                'compute_norm': compute_norm,
+                'data_type': data_type,
+                'acc_type': acc_type,
+                'idx_type': IDX_TYPE,
+                'type_abbrev': type_abbrev,
+                'acc_abbrev': acc_abbrev,
+                'idx_abbrev': IDX_TYPE_ABBREV,
+            })
+
+    return kernels
+
+
+def generate_filename(params):
+    """Generate filename from kernel parameters."""
+    capacity = params['capacity']
+    veclen = params['veclen']
+    ascending = 'true' if params['ascending'] else 'false'
+    compute_norm = 'true' if params['compute_norm'] else 'false'
+    type_abbrev = params['type_abbrev']
+    acc_abbrev = params['acc_abbrev']
+    idx_abbrev = params['idx_abbrev']
+
+    return f"interleaved_scan_kernel_{capacity}_{veclen}_{ascending}_{compute_norm}_{type_abbrev}_{acc_abbrev}_{idx_abbrev}.cu"
+
+
+def generate_cuda_file_content(params):
+    """Generate the content of a CUDA kernel instantiation file."""
+    capacity = params['capacity']
+    veclen = params['veclen']
+    ascending = 'true' if params['ascending'] else 'false'
+    compute_norm = 'true' if params['compute_norm'] else 'false'
+    data_type = params['data_type']
+    acc_type = params['acc_type']
+    idx_type = params['idx_type']
+
+    content = f"""/*
```

Comment:
Sometimes I prefer escaping the first newline to make the embedded code block stay together.
```suggestion
    content = f"""\
/*
```